### PR TITLE
Build: add support for multi arch build

### PR DESCRIPTION
**Pull request by morlay**
**What this PR does**:

add support for multi arch build

split from https://github.com/grafana/tempo/pull/305

**Which issue(s) this PR fixes**:
no fixes

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -35,24 +40,24 @@ endif
...
 .PHONY: tempo-vulture
 tempo-vulture:
-	GO111MODULE=on CGO_ENABLED=0 go build $(GO_OPT) -o ./bin/$(GOOS)/tempo-vulture $(BUILD_INFO) ./cmd/tempo-vulture
+	GO111MODULE=on CGO_ENABLED=0 go build $(GO_OPT) -o ./bin/$(GOOS)/tempo-vulture-$(GOARCH) $(BUILD_INFO) ./cmd/tempo-vulture
 
 .PHONY: exe
 exe:
 	GOOS=linux $(MAKE) $(COMPONENT)
-	cp ./bin/linux/$(COMPONENT) ./cmd/$(COMPONENT)/
+	cp ./bin/linux/$(COMPONENT)-$(GOARCH) ./cmd/$(COMPONENT)/$(COMPONENT)
```
Comment (morlay) [inline Makefile:60]: @joe-elliott is this copied binary used in other place? 

if just for Dockerfile we could remove this. and `rm` in `make docker-component` too.

**Comment by morlay** [comment]
@joe-elliott updates Dockerfile too. I need this for custom buildx build before jaeger-query multi-arch supported.

like https://github.com/querycap/istio

File: Makefile [hunk 2]
```diff
@@ -81,7 +86,7 @@ lint:
 
 .PHONY: docker-component # Not intended to be used directly
 docker-component: check-component exe
-	docker build -t grafana/$(COMPONENT) -f ./cmd/$(COMPONENT)/Dockerfile .
+	docker build -t grafana/$(COMPONENT) --build-arg=TARGETARCH=$(GOARCH) -f ./cmd/$(COMPONENT)/Dockerfile .
```
Comment (morlay) [inline Makefile:89]: pass TARGETARCH here, then dockerfile to pick correctly binary

File: Makefile [hunk 3]
```diff
@@ -4,6 +4,11 @@ VERSION=$(shell ./tools/image-tag | cut -d, -f 1)
 GIT_REVISION := $(shell git rev-parse --short HEAD)
 GIT_BRANCH := $(shell git rev-parse --abbrev-ref HEAD)
 
+NAMESPACE ?= grafana
```
Comment (joe-elliott) [inline Makefile:7]: this is unused.  please remove

**Review by joe-elliott — CHANGES_REQUESTED** [review]
One small nit, but otherwise I like the look of this. 

We can add the ability to use buildx once Jaeger supports multi arch builds.

**Comment by morlay** [comment]
> We can add the ability to use buildx once Jaeger supports multi arch builds.

Dockerfile changes for this goal. 

And should we need to remove the binaries copy?
```
exe:
    cp ./bin/linux/$(COMPONENT)-$(GOARCH) ./cmd/$(COMPONENT)/$(COMPONENT)

docker-component:
    rm ./cmd/$(COMPONENT)/$(COMPONENT)
```

**Comment by morlay** [comment]
@joe-elliott anything need i to do next?

File: Makefile (hunk 1 above)
Comment (joe-elliott) [reply Makefile:58]: I don't believe this binary is used anywhere else.

**Comment by joe-elliott** [comment]
Apologies.  Let this slip.  The headline says "Changes Requested" so I kept looking past it.

So this PR explicitly sets GOOS and GOARCH and gives the user the opportunity to override them?  How does this interact with the Dockerfile?  If we build a linux-arm64 binary and put it in the `alpine:3.9` image is that useful?

**Comment by morlay** [comment]
Since the [oci image](https://github.com/opencontainers/image-spec) lands

`alpine:3.9` is already a multi-arch image (https://hub.docker.com/_/alpine?tab=tags&page=1&name=3.9)

When user run 
```
docker build --platform=linux/arm64 --build-arg=TARGETARCH=arm64
```
docker will pull the `alpine:3.9` image  for `linux/arm64` only and COPY `tempo-arm64` to the final images.

current in Makefile (run on amd64 host)
```bash
docker build --build-arg=TARGETARCH=$(go env GOARCH)
```
docker will pull the `alpine:3.9` image  for `linux/amd64` only and COPY `tempo-amd64` to the final images.
so this will not break current workflow. 

`TARGETARCH` is a [platform arg](https://docs.docker.com/engine/reference/builder/#automatic-platform-args-in-the-global-scope)

in future. we could to buildx, `TARGETARCH` will be automately pass into Docker. it could be just
```bash
docker buildx build --platform=linux/amd64,linux/arm64 --tag=temp
```

equals

```bash
for TARGETPLATFORM in "linux/amd64" "linux/arm64"; do
  docker build --platform=${TARGETPLATFORM} --build-arg=TARGETARCH=${get_arch(TARGETPLATFORM)} --tag=tempo-${TARGETARCH}.
done

docker manifest create tempo tempo-amd64 tempo-arm64 
```

**Comment by morlay** [comment]
no idea about the test failed

```
? 	github.com/grafana/tempo/modules/distributor/receiver	[no test files] 

53panic: runtime error: invalid memory address or nil pointer dereference 

54[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x14f784d] 

55 

56goroutine 426 [running]: 

57github.com/grafana/tempo/tempodb/wal.(*AppendBlock).Find(0x0, 0xc000634538, 0x1, 0x1, 0x1cc0d80, 0xc00032d0e0, 0x0, 0x0, 0x0, 0x0, ...) 

58	/home/runner/work/tempo/tempo/tempodb/wal/append_block.go:143 +0x4d 

59github.com/grafana/tempo/modules/ingester.(*instance).FindTraceByID(0xc00032d0e0, 0xc000634538, 0x1, 0x1, 0x0, 0x0, 0x0) 

60	/home/runner/work/tempo/tempo/modules/ingester/instance.go:227 +0x238 

61github.com/grafana/tempo/modules/ingester.TestInstanceDoesNotRace.func7() 

62	/home/runner/work/tempo/tempo/modules/ingester/instance_test.go:169 +0x92


```

**Comment by morlay** [comment]
@joe-elliott I updates the changes you requested. No idea Why github still say `Changes requested`

**Comment by joe-elliott** [comment]
Thank you for the detailed explanation.  So the final change I would like is to for the `docker build` line to support the 
`--platform=linux/arm64` parameter.

Right now I can override `GOARCH` in the Makefile using an env var, but the --platform param is not correctly set on this line:

https://github.com/grafana/tempo/pull/311/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R86

I will be ready to merge after that!

**Comment by morlay** [comment]
@joe-elliott updated it.

But for final multi-arch builds, prefer to switch to using `buildx` (or using docker build with arch suffix tag, like`latest-amd64`, `latest-arm64`) 
i could create new PR for this until jaeger-query official support multi-arch images.

**Review by joe-elliott — APPROVED** [review]
Excellent.  Thank you for the improvements @morlay!

Agree with your final statement that once Jaeger multi arch builds are in place we can follow suit with buildx.  If you have the bandwidth to tackle this it would be very appreciated.

**Comment by morlay** [comment]
I'm glad to do so. 
I use your cool tools in our multi-arch cluster. 
hope i could merge the multi-arch images [here](https://github.com/querycap/istio) back to related repos.

**Comment by morlay** [comment]
@joe-elliott 

I failed to setup buildx with drone. `--privileged` can't be set in cloud.drone.io.
I don't think drone team could add buildx add `plugins/docker`. https://github.com/drone-plugins/drone-docker/issues/158

if still using drone. have to setup multi-arch image like loki did https://github.com/grafana/loki/blob/master/.drone/drone.yml

**Pull request merged**